pkg/auth/service.go

```diff
@@ -474,456 +539,479 @@ func (s *DBAuthService) ListEffectivePolicies(ctx context.Context, username stri
+	var err error
+	if _, err = s.GetGroup(ctx, groupDisplayName); err != nil {
+		return fmt.Errorf("%s: %w", groupDisplayName, err)
+	}
+
+	// delete user membership to group
+	usersKey := model.KVUserToGroup(groupDisplayName, "")
+	it, err := kv.NewSecondaryIterator(ctx, s.store.Store, (&model.UserData{}).ProtoReflect().Type(), model.PartitionKey, usersKey, "")
+	defer it.Close()
+	if err != nil {
+		return err
+	}
+	for it.Next() {
+		entry := it.Entry()
+		user := entry.Value.(*model.UserData)
+		if err = s.RemoveUserFromGroup(ctx, user.Username, groupDisplayName); err != nil {
+			return fmt.Errorf("%s: %w", user.Username, err)
+		}
+	}
+	if err = it.Err(); err != nil {
+		return err
+	}
+
+	// delete policy attachment to group
+	policiesKey := model.KVPolicyToGroup(groupDisplayName, "")
+	itr, err := kv.NewSecondaryIterator(ctx, s.store.Store, (&model.PolicyData{}).ProtoReflect().Type(), model.PartitionKey, policiesKey, "")
+	defer it.Close()
+	if err != nil {
+		return err
+	}
+	for itr.Next() {
+		entry := itr.Entry()
+		policy := entry.Value.(*model.PolicyData)
+		if err = s.DetachPolicyFromGroup(ctx, policy.DisplayName, groupDisplayName); err != nil {
+			return fmt.Errorf("%s: %w", policy.DisplayName, err)
+		}
+	}
+	if err = itr.Err(); err != nil {
+		return err
+	}
+
+	// delete group
+	groupPath := model.KVGroupPath(groupDisplayName)
+	err = s.store.DeleteMsg(ctx, model.PartitionKey, groupPath)
+	if err != nil {
+		return fmt.Errorf("delete user (userKey %s): %w", groupPath, err)
+	}
 	return err
 }
 
-func (s *DBAuthService) GetGroup(ctx context.Context, groupDisplayName string) (*model.Group, error) {
-	group, err := s.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		return getGroup(tx, groupDisplayName)
-	}, db.ReadOnly())
+func (s *KVAuthService) GetGroup(ctx context.Context, groupDisplayName string) (*model.Group, error) {
+	groupKey := model.KVGroupPath(groupDisplayName)
+	m := model.GroupData{}
+	_, err := s.store.GetMsg(ctx, model.PartitionKey, groupKey, &m)
 	if err != nil {
+		if errors.Is(err, kv.ErrNotFound) {
+			err = ErrNotFound
+		}
 		return nil, err
 	}
-	return model.ConvertGroup(group.(*model.DBGroup)), nil
+	return model.GroupFromProto(&m), nil
 }
 
-func (s *DBAuthService) ListGroups(ctx context.Context, params *model.PaginationParams) ([]*model.Group, *model.Paginator, error) {
-	var group model.DBGroup
-	slice, paginator, err := ListPaged(ctx, s.db, reflect.TypeOf(group), params, "display_name",
-		psql.Select("*").
-			From("auth_groups").
-			Where(sq.Like{"display_name": fmt.Sprint(params.Prefix, "%")}))
-	if err != nil {
+func (s *KVAuthService) ListGroups(ctx context.Context, params *model.PaginationParams) ([]*model.Group, *model.Paginator, error) {
+	var group model.GroupData
+	groupKey := model.KVGroupPath("")
+
+	msgs, paginator, err := s.ListKVPaged(ctx, (&group).ProtoReflect().Type(), params, groupKey, false)
```

Comment:
ListKVPaged uses `PrefixIterator` with `groupKey` as a prefix.
The value assigned to `groupKey` is `groups/` which is also the prefix to all user membership in group entries - `groups/<DisplayName>/users/<Username>` - and all group <--> policy relations - `groups/<Displayname>/policies/<DisplayName>`
The result slice contains a lot more than only groups
One option is to change the keys schema so it would be easy to distinguish between the entities (i.e. keys that cannot be prefixes of each other).
Another (ugly) option is to filter out irrelevant result as you iterate - not sure how trivial it is.

Reply:
I believe the same goes to all similar cases, e.g. `ListUsers` and the correlated user <--> relations

Reply:
done

Reply:
Great. Thanks